### PR TITLE
RFC: Authorization fallback

### LINKED ISSUE #1906: Error handling on the authorization request
My understanding is that if the authorization request results in an error response (4xx or 5xx), the page will render without any AMP Access expression processing.

First it would be good to state this in the documentation, and make it clear that expressions will simply not be evaluated (assuming that is the behaviour), so in the example below the section would be hidden on error.

```
<section amp-access="NOT forbidden" amp-access-hide>
  THIS IS PREMIUM CONTENT
</section>
```

Second, it would be good to provide some way for developers to detect when an error occurs, because they might want to make a decision whether to fail open or fail closed, or provide some message to the user. I can think of a couple of ways to achieve this:

- Provide a 'fallback' response JSON which will be used if the authorization request fails. This has the advantage that developers can vary this JSON based on business rules, such as whether the content is premium or not, without changing the rest of the page logic
- Provide a new attribute like amp-access-error which will only show the content when an error occurs

### PR DESCRIPTION
**Pull request by dvoytenko**
Closes #1906.

(No tests yet)

This RFC item proposes to add an optional "authorizationFallback" configuration property that will be used in case when the authorization fails.

/cc @jugglingcats, @ashwinlimaye, @sebastianbenz, @andreban

### CODE REVIEW

**Comment by dvoytenko** [comment]
@cramforce This is now ready for review. PTAL.

File: extensions/amp-access/0.1/amp-access.js [hunk 1]
```diff
@@ -350,6 +352,16 @@ export class AccessService {
             credentials: 'include',
             requireAmpResponseSourceOrigin: true
           }));
+    }).catch(error => {
+      this.analyticsEvent_('access-authorization-failed');
```
Comment (cramforce) [inline extensions/amp-access/0.1/amp-access.js]: Please write a test for this.

File: extensions/amp-access/0.1/amp-access.js [hunk 2]
```diff
@@ -350,6 +352,16 @@ export class AccessService {
...
+      if (this.config_.authorizationFallback) {
+        // Use fallback.
+        log.error(TAG, 'Authorization failed, will use fallback: ', error);
```
Comment (cramforce) [inline extensions/amp-access/0.1/amp-access.js]: Dev visible warn?

File: examples/article-access.amp.html [hunk 3]
```diff
@@ -9,7 +9,8 @@
     {
       "authorization": "http://localhost:8002/amp-authorization.json?rid=READER_ID&url=CANONICAL_URL&_=RANDOM",
       "pingback": "http://localhost:8002/amp-pingback?rid=READER_ID&url=CANONICAL_URL&views=AUTHDATA(views)",
-      "login": "http://localhost:8002/amp-login?rid=READER_ID&url=CANONICAL_URL"
+      "login": "http://localhost:8002/amp-login?rid=READER_ID&url=CANONICAL_URL",
+      "authorizationFallback": {"error": true}
```
Comment (cramforce) [inline examples/article-access.amp.html]: Maybe `…FallbackResponse`?

File: extensions/amp-access/0.1/amp-access.js (hunk 1 above)
Comment (cramforce) [inline extensions/amp-access/0.1/amp-access.js]: I mean the full function

File: extensions/amp-access/0.1/amp-access.js [hunk 4]
```diff
@@ -350,6 +352,16 @@ export class AccessService {
...
+        return this.config_.authorizationFallback;
```
Comment (cramforce) [inline extensions/amp-access/0.1/amp-access.js]: Maybe rethrow in timeout for reporting?

File: extensions/amp-access/0.1/amp-access.js [hunk 5]
```diff
@@ -350,6 +352,16 @@ export class AccessService {
             credentials: 'include',
             requireAmpResponseSourceOrigin: true
           }));
+    }).catch(error => {
```
Comment (cramforce) [inline extensions/amp-access/0.1/amp-access.js]: File bug for retry?

File: extensions/amp-access/0.1/amp-access.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/amp-access.js]: Yes. Working on tests still as specified above. Will be available soon.

File: extensions/amp-access/0.1/amp-access.js (hunk 2 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/amp-access.js]: I switched to rethrow the error.

File: examples/article-access.amp.html (hunk 3 above)
Comment (dvoytenko) [inline examples/article-access.amp.html]: Sounds good. It's not `authorizationFallbackResponse`. A bit wordy, but at least it's unambiguous.

File: extensions/amp-access/0.1/amp-access.js (hunk 4 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/amp-access.js]: Doe

File: extensions/amp-access/0.1/amp-access.js (hunk 5 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/amp-access.js]: I'd rather steer clear of retry initially - that would increase load on target servers.

File: extensions/amp-access/0.1/amp-access.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/amp-access.js]: Done

**Comment by dvoytenko** [comment]
@cramforce PTAL.

File: extensions/amp-access/0.1/amp-access.js (hunk 5 above)
Comment (cramforce) [inline extensions/amp-access/0.1/amp-access.js]: No prob. We already have a primitive for exponential backoff in place. But I would not retry more than once anyway.

**Comment by cramforce** [comment]
LGTM
